### PR TITLE
LSP: Push diagnostics collected on project indexing.

### PR DESCRIPTION
**Pull request by dbalek**
LSP server should notify its client to publish diagnostics collected on project scan.

### CODE REVIEW

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/ErrorsNotifier.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+            servers.keySet().removeAll(toRemove);
+        }
+        try {
+            Collection<? extends URL> filesWithErrors = ErrorsCache.getAllFilesWithRecord(root);
+            if (!filesWithErrors.isEmpty()) {
+                Project project = FileOwnerQuery.getOwner(root.toURI());
+                for (LspServerState server : toProcess) {
+                    for (Project p : server.openedProjects().getNow(new Project[0])) {
+                        if (p == project) {
+                            Diagnostic.ReporterControl control = ((TextDocumentServiceImpl)server.getTextDocumentService()).createReporterControl();
```
Comment (sdedic) [inline java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/ErrorsNotifier.java:71]: Nitpick: might be less tangled to the implementatation: `Diagnostic.findReporterControl(Lookup.singleton(server), null)`

**Review by sdedic — APPROVED** [review]

**Comment by lahodaj** [comment]
I may be doing something wrong or misunderstanding something, but a few comments/questions:
- when I try currently, I don't seem to get the effects for Java errors, but I am getting this in the log:
```
java.lang.Exception: no NbCodeLanguageClient!
	at org.netbeans.modules.java.lsp.server.protocol.TextDocumentServiceImpl.runDiagnosticTasks(TextDocumentServiceImpl.java:1890)
	at org.netbeans.modules.java.lsp.server.protocol.TextDocumentServiceImpl.runDiagnosticTasks(TextDocumentServiceImpl.java:1873)
	at org.netbeans.modules.java.lsp.server.protocol.TextDocumentServiceImpl.access$500(TextDocumentServiceImpl.java:267)
	at org.netbeans.modules.java.lsp.server.protocol.TextDocumentServiceImpl$5.lambda$diagnosticChanged$0(TextDocumentServiceImpl.java:2179)
	at java.base/java.util.concurrent.CompletableFuture.uniRunNow(CompletableFuture.java:819)
	at java.base/java.util.concurrent.CompletableFuture.uniRunStage(CompletableFuture.java:803)
	at java.base/java.util.concurrent.CompletableFuture.thenRun(CompletableFuture.java:2227)
	at org.netbeans.modules.java.lsp.server.protocol.TextDocumentServiceImpl$5.diagnosticChanged(TextDocumentServiceImpl.java:2178)
	at org.netbeans.modules.java.lsp.server.protocol.ErrorsNotifier.notifyErrors(ErrorsNotifier.java:73)
	at org.netbeans.modules.java.lsp.server.protocol.Server$CustomIndexerTelemetryFactory.scanFinished(Server.java:1416)
	at org.netbeans.modules.parsing.impl.indexing.RepositoryUpdater$Work.scanFinished(RepositoryUpdater.java:2518)
...
```

- the `ErrorsNotifier` seems to refresh diagnostics only for files for which there's an entry in the index. But what if there was an error in the index, and the error is fixed? As I read the code, the diagnostics will not be refreshed?

- what if there will be a lot of files with errors? Will this try to open the Document for all of them, and do full diagnostic run for them, parsing&attributing in the process?

**Review by MartinBalin — APPROVED** [review]

File: enterprise/micronaut/src/org/netbeans/modules/micronaut/hints/MicronautConfigErrorProvider.java [hunk 2]
```diff
@@ -221,4 +196,19 @@ public int getIndexVersion() {
...
     }
+
+    private static final class ErrorConvertorImpl implements ErrorsCache.Convertor<Diagnostic> {
+        @Override
+        public ErrorsCache.ErrorKind getKind(Diagnostic t) {
+            return t.getSeverity() == Diagnostic.Severity.Error ? ErrorsCache.ErrorKind.ERROR : ErrorsCache.ErrorKind.WARNING;
+        }
+        @Override
+        public int getLineNumber(Diagnostic t) {
+            return Integer.parseInt(t.getCode().substring(26));
```
[the remark this answers is not in the document]
Comment (dbalek) [reply enterprise/micronaut/src/org/netbeans/modules/micronaut/hints/MicronautConfigErrorProvider.java]: Fixed.

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/ErrorsNotifier.java (hunk 1 above)
Comment (dbalek) [reply java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/ErrorsNotifier.java]: Fixed.

**Comment by dbalek** [comment]
>     * when I try currently, I don't seem to get the effects for Java errors, but I am getting this in the log:

Fixed now. Thanks for pointing this out.

>     * the `ErrorsNotifier` seems to refresh diagnostics only for files for which there's an entry in the index. But what if there was an error in the index, and the error is fixed? As I read the code, the diagnostics will not be refreshed?

Currently - the refresh is done in the following way: For the file being modified, the diagnostics are updated with each modification via LSP ‘textDocument/diagnostic’ requests. After the modified document is saved - it is re-indexed (`ErrorsNotifier` gets started and goes for all files in index with errors and recompute their diagnostic). After that the second round of re-indexing starts with all dependent files and files with error (which removes errors from the index for files that were transitively fixed with the change in the original file).
 
>     * what if there will be a lot of files with errors? Will this try to open the Document for all of them, and do full diagnostic run for them, parsing&attributing in the process?

You are right - currently it does the full diagnostic run for all files marked as erroneous including parsing&attributing in the process. If it turns out to harm the overall performance of the extension, it can be rewritten to publish the diagnostic directly form the caches.

**Pull request merged**